contracts/base/ERC721Base.sol

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import { ERC721A } from "../eip/ERC721A.sol";
+
+import "../extension/ContractMetadata.sol";
+import "../extension/Multicall.sol";
+import "../extension/Ownable.sol";
+import "../extension/RoyaltyUpdated.sol";
+import "../extension/BatchMintMetadata.sol";
+
+import "../lib/TWStrings.sol";
+
+/**
+ *  The `ERC721Base` smart contract implements the ERC721 NFT standard, along with the ERC721A optimization to the standard.
+ *  It includes the following additions to standard ERC721 logic:
+ *
+ *      - Ability to mint NFTs via the provided `mint` function.
+ *
+ *      - Contract metadata for royalty support on platforms such as OpenSea that use
+ *        off-chain information to distribute roaylties.
+ *
+ *      - Ownership of the contract, with the ability to restrict certain functions to
+ *        only be called by the contract's owner.
+ *
+ *      - Multicall capability for fetching NFT data.
+ *
+ *      - EIP 2981 compliance for outright royalty support on NFT marketplaces.
+ */
+
+contract ERC721Base is ERC721A, ContractMetadata, Multicall, Ownable, Royalty, BatchMintMetadata {
+    using TWStrings for uint256;
+
+    /*//////////////////////////////////////////////////////////////
+                            Constructor
+    //////////////////////////////////////////////////////////////*/
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        string memory _contractURI,
+        address _royaltyRecipient,
+        uint128 _royaltyBps
```

Comment:
what's the reasoning for using `_variable` vs just `variable` ? is there a convention?

Reply:
Yes. (1) Largely to prevent shadowing state variables with he same name, and (2) as a visual rule of thumb; if `_` is present in the name, it's an argument and should be validated via require checks etc., and if it's not present, it's a state / in-memory variable.